src/collections/skip_list/mod.rs

```diff
@@ -0,0 +1,449 @@
+    }
+    /// Removes the entry from the list.
+    // pub fn remove(&mut self, key: &K) -> Option<Entry<K, V>> {
+    //     let mut cached: Vec<*mut SkipNode<K, V>> = Vec::new();
+    //     let mut start_point = self.head;
+    //     for level in (0..MAX_HEIGHT).rev() {
+    //         unsafe {
+    //             while let Some(next_node) = (*start_point).next_by_height[level] {
+    //                 if (*next_node).entry.key == *key {
+
+    //                 }
+    //             }
+    //         }
+    //     }
+    //     None;
+    // }
+    /// Remove the first entry under key order.
+    pub fn pop_first(&mut self) -> Option<Entry<K, V>> {
+        let head_node = unsafe { &mut *self.head };
+        head_node.next_by_height[0].take().map(|first_node_ptr| {
+            let mut first_node = unsafe { Box::from_raw(first_node_ptr) };
+            for (level, transmit_ptr) in first_node.next_by_height.iter_mut().enumerate() {
+                head_node.next_by_height[level] = transmit_ptr.take();
+            }
+            self.len -= 1;
+            first_node.entry
+        })
+    }
+    /// Creates a consuming iterator yielding key-value tuple in specifi order.
+    pub fn into_iter(self) -> IntoIterator<K, V> {
+        IntoIterator { list: self }
+    }
+    /// Creates an anonymous iterator yields the key in specifi order.
+    pub fn keys(&self) -> Box<dyn Iterator<Item = &K> + '_> {
+        Box::new(self.iter().map(|(key, _)| key)) as Box<dyn Iterator<Item = &K>>
+    }
+    /// Creates an anonymous iterator yields the values in specifi order.
+    pub fn values(&self) -> Box<dyn Iterator<Item = &V> + '_> {
+        Box::new(self.iter().map(|(_, value)| value)) as Box<dyn Iterator<Item = &V>>
+    }
+    /// Creates an anonymous iterator yields the values in range [key1, key2).
+    pub fn range(&self, key1: K, key2: K) -> Box<dyn Iterator<Item = (&K, &V)> + '_> {
+        Box::new(
+            self.iter()
+                .filter(move |(key, _)| (*key >= &key1) && (*key < &key2)),
+        )
+    }
+    /// Indiates whether the skiplist is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+    /// Gets an iterator over the list, sorted by key.
+    pub fn iter(&self) -> Iter<K, V> {
+        Iter {
+            ptr: unsafe { (*self.head).next_by_height[0] },
+            _marker: Default::default(),
+        }
+    }
+    /// Gets an mutable iterator over the list, sorted by key.
+    pub fn iter_mut(&mut self) -> MutIterator<K, V> {
+        MutIterator {
+            ptr: unsafe { (*self.head).next_by_height[0] },
+            _marker: Default::default(),
+        }
+    }
+}
+
+impl<K, V> Display for SkipList<K, V>
+where
+    K: Default,
+    V: Default,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "SkipList:\n {:?}", self.head)
+    }
+}
+
+impl<K, V> Drop for SkipList<K, V>
+where
+    K: Default,
+    V: Default,
+{
+    fn drop(&mut self) {
```

Comment:
Have only skimmed unsafe part but found that something drops twice unexpectly. You can add a test to check dropping count.

```rust
#[test]
fn drop() {
    static mut DROPS: u32 = 0;

    #[derive(Default)]
    struct S;

    impl Drop for S {
        fn drop(&mut self) {
            unsafe {
                DROPS += 1;
            }
        }
    }

    let count = 10;
    let mut d = SkipList::new();
    (0..count).for_each(|k| d.insert(k, S));

    core::mem::drop(d);

    unsafe {
        assert_eq!(DROPS, count);
    }
}
```

Reply:
Memory error is hard for me noob, I will try turn to the Safe Rust.